# Тема “Обучение с учителем”

### Задание 1

Импортируйте библиотеки pandas и numpy.

In [2]:
import numpy as np
import pandas as pd
import os

Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.

In [3]:
from sklearn.datasets import load_boston

In [4]:
boston = load_boston()

In [5]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [6]:
data = boston.data

In [7]:
data.shape

(506, 13)

In [8]:
data

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

target - цена, которую нужно предсказать

In [9]:
target = boston.target

In [10]:
feature_names = boston.feature_names

In [11]:
feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [13]:
for line in boston.DESCR.split('\n'):
    print(line)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

In [14]:
X = pd.DataFrame(data, columns=feature_names)

In [15]:
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [16]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
CRIM       506 non-null float64
ZN         506 non-null float64
INDUS      506 non-null float64
CHAS       506 non-null float64
NOX        506 non-null float64
RM         506 non-null float64
AGE        506 non-null float64
DIS        506 non-null float64
RAD        506 non-null float64
TAX        506 non-null float64
PTRATIO    506 non-null float64
B          506 non-null float64
LSTAT      506 non-null float64
dtypes: float64(13)
memory usage: 51.5 KB


In [17]:
y = pd.DataFrame(target, columns=['price'])

In [18]:
y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

In [19]:
y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 1 columns):
price    506 non-null float64
dtypes: float64(1)
memory usage: 4.0 KB


Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.

In [20]:
from sklearn.model_selection import train_test_split

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [29]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 152 entries, 173 to 447
Data columns (total 13 columns):
CRIM       152 non-null float64
ZN         152 non-null float64
INDUS      152 non-null float64
CHAS       152 non-null float64
NOX        152 non-null float64
RM         152 non-null float64
AGE        152 non-null float64
DIS        152 non-null float64
RAD        152 non-null float64
TAX        152 non-null float64
PTRATIO    152 non-null float64
B          152 non-null float64
LSTAT      152 non-null float64
dtypes: float64(13)
memory usage: 16.6 KB


Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

In [30]:
from sklearn.linear_model import LinearRegression

In [32]:
lr = LinearRegression()

Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

In [34]:
lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [35]:
y_pred = lr.predict(X_test)

In [36]:
check_test = pd.DataFrame({'y_test': y_test['price'],
                           'y_pred': y_pred.flatten()},
                          columns=['y_test', 'y_pred'])

In [37]:
check_test.head(10)

y_test     y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280
76     20.0  23.146689
316    17.8  17.392124
140    14.0  14.078599
471    19.6  23.036927
500    16.8  20.599433

Средняя квадратичная ошибка

In [39]:
check_test['error'] = check_test['y_pred'] - check_test['y_test']

In [40]:
check_test.head(10)

y_test     y_pred     error
173    23.6  28.648960  5.048960
274    32.4  36.495014  4.095014
491    13.6  15.411193  1.811193
72     22.8  25.403213  2.603213
452    16.1  18.855280  2.755280
76     20.0  23.146689  3.146689
316    17.8  17.392124 -0.407876
140    14.0  14.078599  0.078599
471    19.6  23.036927  3.436927
500    16.8  20.599433  3.799433

In [41]:
initial_mse = (check_test['error'] ** 2).mean()
initial_mse

21.517444231177

In [42]:
from sklearn.metrics import mean_squared_error

In [43]:
initial_mse = mean_squared_error(y_test, y_pred)
initial_mse

21.517444231176995

Средняя абсолютная ошибка - не так сильно реагирует на выбросы данных

In [44]:
(np.abs(check_test['error'])).mean()

3.162709871457379

In [45]:
from sklearn.metrics import mean_absolute_error

In [46]:
mean_absolute_error(y_test, y_pred)

3.1627098714573783

Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

R2 - не зависит от масштабов предсказываемой величины, её значение не превышает 1

In [48]:
from sklearn.metrics import r2_score

In [49]:
r2_score(y_test, y_pred)

0.711226005748496

Просмотр коэффициентов линейной регрессии

w0

In [50]:
lr.intercept_

array([31.63108404])

w1...

In [52]:
lr.coef_

array([[-1.33470103e-01,  3.58089136e-02,  4.95226452e-02,
         3.11983512e+00, -1.54170609e+01,  4.05719923e+00,
        -1.08208352e-02, -1.38599824e+00,  2.42727340e-01,
        -8.70223437e-03, -9.10685208e-01,  1.17941159e-02,
        -5.47113313e-01]])

In [53]:
import matplotlib.pyplot as plt
%matplotlib inline
%config InlineBackend.figure_format = 'svg'

In [ ]:
plt.barh(X_train.columns, lr.coef_.flatten())
plt.xlabel('Вес признака')
plt.ylabel('Признак')

### Задание 2

Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42.

In [56]:
from sklearn.ensemble import RandomForestRegressor

In [58]:
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy, так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.

In [60]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=12,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=1000,
                      n_jobs=None, oob_score=False, random_state=42, verbose=0,
                      warm_start=False)

Сделайте предсказание на тестовых данных и посчитайте R2.

In [61]:
y_pred = model.predict(X_test)

In [62]:
check_test = pd.DataFrame({'y_test': y_test['price'],
                           'y_pred': y_pred.flatten()},
                          columns=['y_test', 'y_pred'])

In [63]:
check_test.head(10)

y_test     y_pred
173    23.6  22.846138
274    32.4  31.156114
491    13.6  16.297226
72     22.8  23.821036
452    16.1  17.212148
76     20.0  21.820092
316    17.8  19.866369
140    14.0  14.759938
471    19.6  21.235224
500    16.8  20.883103

In [64]:
r2_score(y_test, y_pred)

0.8749965273218174

Сравните с результатом из предыдущего задания.

ОТВЕТ: Значение R2 ближе к 1 (0.8749965273218174 по сравнению с 0.711226005748496), следовательно предсказание сделано более точно.

Напишите в комментариях к коду, какая модель в данном случае работает лучше.

ОТВЕТ: В данном случае лучше работает модель ForestRegressor.

In [83]:
?RandomForestRegressor

In [70]:
model.feature_importances_

array([0.03211748, 0.00154999, 0.0070941 , 0.0011488 , 0.01436832,
       0.40270459, 0.01424477, 0.06403265, 0.00496762, 0.01169177,
       0.01808961, 0.0123114 , 0.41567892])

In [72]:
sum(model.feature_importances_)

1.0

In [80]:
print(sorted(model.feature_importances_)[-2:])

[0.402704591696731, 0.4156789214509943]


Если я правильно понял, в данной модели строится "случайный лес" и у этих признаков нет каких-то специфических названий, все прихнаки созданы случайным образом и два вышеуказанных наиболее важны. То есть обратиться к ним можно по их номеру...???

К сожалению, на 4-ое задание не хватило времени. :-(